packages/peregrine/lib/talons/CheckoutPage/useShippingMethod.js

```diff
@@ -61,10 +64,15 @@ export const useShippingMethod = props => {
     /*
      *  State / Derived state.
      */
-    const [displayState, setDisplayState] = useState(displayStates.EDITING);
-    const [shippingMethods, setShippingMethods] = useState([]);
-    const [selectedShippingMethod, setSelectedShippingMethod] = useState(null);
+    const [displayState, setDisplayState] = useState(
+        displayStates.INITIALIZING
+    );
+    const [isBackgroundAutoSelecting, setIsBackgroundAutoSelecting] = useState(
+        false
+    );
     const [isUpdateMode, setIsUpdateMode] = useState(false);
+    const [selectedShippingMethod, setSelectedShippingMethod] = useState(null);
+    const [shippingMethods, setShippingMethods] = useState([]);
```

Comment:
I know this pattern existing prior to this PR, but wanted to discuss why I think it might be an anti-pattern, and something we should avoid. Apollo's hooks are effects themselves, so you should never need to wrap that returned state in an effect, or duplicate it into its own state. I'm a bit unique in that I didn't write much React before hooks, but I think the pattern you're going for here from previous paradigms is derivedState.

Two use cases for derived state:
1. Data needs manipulated - `useMemo` if complex, but option 2 also works, wrap logic in a function:
```js
const memoizedShippingMethods = useMemo(() => {
    let mutatedShippingMethods = [];
    if (data) {
        const primaryShippingAddress = data.cart.shipping_addresses[0];

        // Shape the list of available shipping methods.
        // Sort them by price and add a serialized property to each.
        const rawShippingMethods =
            primaryShippingAddress.available_shipping_methods;
        const shippingMethodsByPrice = [...rawShippingMethods].sort(
            byPrice
        );
        mutatedShippingMethods = shippingMethodsByPrice.map(
            addSerializedProperty
        );
    }

    return mutatedShippingMethods;
}, [data]);
```

2. Data just needs extracted - no special treatment
```js
const derviedSelectedShippingMethod =
    data &&
    addSerializedProperty(
        data.cart.shipping_addresses[0].selected_shipping_method
    ); // you may want a || null default value here
```

I would like to earmark this to be discussed during tomorrow's team time, as I think we should make sure we're all aligned on this. I'm not 100% I'm right, so worth discussing for my own benefit as well.

Reply:
We use the derived pattern often throughout the codebase. [Here](https://github.com/magento/pwa-studio/blob/52fcaafab570dd34f3a42620defd96bf7feb2385/packages/peregrine/lib/talons/CheckoutPage/ShippingInformation/useShippingInformation.js#L31-L59) it is in the shipping info talon and [here](https://github.com/magento/pwa-studio/blob/52fcaafab570dd34f3a42620defd96bf7feb2385/packages/peregrine/lib/talons/Breadcrumbs/useBreadcrumbs.js#L42-L54) it is in the breadcrumb talon.

Efficiency aside I think it's more readable which is why I'd prefer it :)